src/redash_stmo/datasource_health/health.py

```diff
@@ -0,0 +1,92 @@
+from redash.worker import celery
+from redash.utils import parse_human_time
+from redash.monitor import get_status
+from redash.query_runner import BaseQueryRunner
+
+original_get_status = get_status
+logger = get_task_logger(__name__)
+
+
+def update_health_status(data_source_id, data_source_name, query_text, data):
+    key = "data_sources:health"
+
+    cache = json.loads(redis_connection.get(key) or '{}')
+    if data_source_id not in cache:
+        cache[data_source_id] = {
+            "metadata": { "name": data_source_name },
+            "queries": {}
+        }
+    cache[data_source_id]["queries"][query_text] = data
+
+    cache[data_source_id]["status"] = "SUCCESS"
+    for query_status in cache[data_source_id]["queries"].values():
+        if query_status["status"] == "FAIL":
+            cache[data_source_id]["status"] = "FAIL"
+            break
+
+    redis_connection.set(key, json.dumps(cache))
+
+
+@celery.task(name="redash_stmo.datasource_health.health.health_status", time_limit=90, soft_time_limit=60)
+def health_status():
+    for ds in models.DataSource.query:
+        logger.info(u"task=health_status state=start ds_id=%s", ds.id)
+
+        runtime = None
+        query_text = ds.query_runner.noop_query
+        custom_queries = settings.CUSTOM_HEALTH_QUERIES
+        ds_id = str(ds.id)
+
+        if custom_queries and ds_id in custom_queries:
+            query_text = custom_queries[ds_id]
+
+        try:
+            start_time = time.time()
+            ds.query_runner.test_connection_extended(query_text)
+            runtime = time.time() - start_time
+        except Exception as e:
+            logger.warning(u"Failed health check for the data source: %s", ds.name, exc_info=1)
+            statsd_client.incr('health_status.error')
+            logger.info(u"task=health_status state=error ds_id=%s runtime=%.2f", ds.id, time.time() - start_time)
+
+        update_health_status(ds_id, ds.name, query_text, {
+            "status": "SUCCESS" if runtime is not None else "FAIL",
+            "last_run": start_time,
+            "last_run_human": str(parse_human_time(str(start_time))),
+            "runtime": runtime
+        })
+
+def test_connection(self, custom_query_text=None):
+    if self.noop_query is None:
+        raise NotImplementedError()
+
+    query_text = custom_query_text or self.noop_query
+    data, error = self.run_query(query_text, None)
+
+    if error is not None:
+        raise Exception(error)
+
+def custom_get_status():
+    status = original_get_status()
+    status['data_sources'] = json.loads(redis_connection.get('data_sources:health') or '{}')
+    return status
+
+def task_info(app=None):
+    # Overriding a couple of functions
+    BaseQueryRunner.test_connection_extended = test_connection
+    redash.handlers.get_status = custom_get_status
+
+    task_info = {
+        "task": health_status,
+        "interval_in_seconds": settings.HEALTH_QUERIES_REFRESH_SCHEDULE
+    }
```

Comment:
I believe this is a leftover from your previous plan to have task entrypoints that are automatically registered. I'm sorry I didn't go into more detail after the last meeting we talked about, the pattern I thought of is the one [Redash uses for enabling periodic tasks](https://github.com/getredash/redash/blob/f78ab06369fec47abe0856c53ee189a12f3d5c29/redash/settings/__init__.py#L38-L41).

Basically in the [`redash/worker.py`](https://github.com/getredash/redash/blob/f78ab06369fec47abe0856c53ee189a12f3d5c29/redash/worker.py#L41-L45) file it'll check the settings whether it should add that specific task to the schedule.

What I propose is that we use the same pattern to define a similar setting in `redash_stmo/settings.py` and then update the celery config on demand in the entrypoint callback, basically auto-registering the task when Flask starts up, e.g.:

```python
from redash.worker import celery_schedule
from . import settings

if settings.HEALTH_STATUS_UPDATE_ENABLED:
    celery_schedule['update_health_status'] = {
        'task': 'redash_stmo.health.update_health_status',  # the Celery task
        'schedule': timedelta(minutes=settings.HEALTH_QUERIES_REFRESH_SCHEDULE)
    }
    celery.conf.update(CELERYBEAT_SCHEDULE=celery_schedule)
```

Reply:
Actually, this was the pattern I initially tried to use, but unfortunately it doesn't seem to be working. I did some reading and my understanding is that `add_periodic_task()` (http://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html#entries) can be used to do this, however it needs to have a `@app.on_after_configure.connect` decorator in order to be processed at any point after initial configuration has been complete. Since `add_periodic_task()` sets `celery.conf.beat_schedule` in the background, we would also need to update `celery.conf.beat_schedule` under the `@app.on_after_configure.connect` decorator. However, this decorator seems to have no effect when used here in `health.py`. I suspect this is because the extensions get initialized too late, after the `on_after_configure` event has already fired and been missed.

Additionally, there seemed to be some stackoverflow comments on this here:
https://stackoverflow.com/questions/41124591/setting-up-periodic-tasks-in-celery-celerybeat-dynamically-using-add-periodic


> In order to be able to dynamically manage periodic tasks and reschedule celerybeat at runtime, there’s the django-celery-beat extension that stores the schedule in the Django database, and presents a convenient admin interface to manage periodic tasks at runtime.

Which sounds to me like dynamic celery beat schedule updating only works with the django extension.

This is why I ended up with the design you see here. And on the redash side, I did this: https://github.com/mozilla/redash/pull/444/files#diff-ad93995bb355251c02f87fc02db69351R82

This was the only way I could get it to work. Please let me know if you see something I've missed

Reply:
Aah, I should have probably have seen this, since in fact ATMO had the same problems of having the schedule tasks dynamically and in the end I used [redbeat](https://github.com/sibson/redbeat) for that (https://github.com/mozilla/telemetry-analysis-service/pull/318 and https://github.com/mozilla/telemetry-analysis-service/pull/379), which is a similar thing like the old, now defunct django-celery package. It basically provides a custom scheduler that persists the periodic task list in Redis and a bit of API to add/remove it after startup.

That said I don't think the use cases are the same and if we can solve this with the MUCH smaller solution of yours let's do it.

In Flask the best practice for extensions is to have them [register themeselves](http://flask.pocoo.org/docs/1.0/api/#flask.Flask.extensions). So I would suggest to change the code in https://github.com/mozilla/redash/pull/444/files#diff-ad93995bb355251c02f87fc02db69351R82 to not populate the list there but do it in the extension itself, e.g.::

```python
def datasource_health(app):
    """A Redash extension to add datasource health status reporting."""

    # Override the default status API view with our extended view
    app.view_functions['%s.status_api' % routes.name] = stmo_status_api
 
    # Add the update_health_status task to a list of periodic tasks
    if not hasattr(app, 'periodic_tasks'):
        app.periodic_tasks = {}

    app.periodic_tasks['update_health_status'] = {
        'signature': update_health_status.s(),
        'schedule': settings.HEALTH_QUERIES_REFRESH_SCHEDULE,
    }
```

As you can see, instead of inventing our own return format, I've just used a dictionary of keyword parameters that [add_periodic_task expects](https://github.com/celery/celery/blob/eeda18611ceed2560145f95ada4977a1b825d282/celery/app/base.py#L974-L975) so we keep the design open for any kind of tasks that Celery supports and other Redash extensions may need in the future.

The use of `.s()` for the signature parameter is also significant since it'll allow Celery to store information about a task in its task store as metadata instead of as Python callables, which are hard to serialize for example to JSON (the recently new default in Celery). In other words, this will be a great help to stay consistent with Celery API design and by that reduce risk for using tasks like this.